anesthetic/utils.py

```diff
@@ -89,6 +90,148 @@ def quantile(a, q, w=None, interpolation='linear'):
     return quant
 
 
+def sample_cdf(samples, inverse=False, interpolation='linear'):
+    """Sample the empirical cdf for a 1d array."""
+    samples = np.array(samples)
+    samples.sort()
+    ngaps = len(samples)-1
+    gaps = np.random.dirichlet(np.ones(ngaps))
+    cdf = np.array([0, *np.cumsum(gaps)])
+    assert np.isclose(cdf[-1], 1, atol=1e-9, rtol=1e-9), \
+        "Error: CDF does not reach 1 but "+str(cdf[-1])
+    # Set the last element (tested to be approx 1)
+    # to exactly 1 to avoid interpolation errors
+    cdf[-1] = 1
+    if inverse:
+        return interp1d(cdf, samples, kind=interpolation)
+    else:
+        return interp1d(samples, cdf, kind=interpolation)
+
+
+def credibility_interval(samples, weights=None, level=0.68, method="hpd",
+                         u=None, n_iter=12, return_covariance=False,
+                         verbose=False):
+    """Compute the credibility interval of weighted samples.
+
+    Based on linear interpolation of the cumulative density function, thus
+    expect discretization errors on the scale of distances between samples.
+
+    https://github.com/Stefan-Heimersheim/fastCI#readme
+
+    Parameters
+    ----------
+    samples : array
+        Samples to compute the credibility interval of.
+    weights : array, default=np.ones_like(samples)
+        Weights corresponding to samples.
+    level : float, default=0.68
+        Credibility level (probability, <1).
+    method : str, default='hpd'
+        Which definition of interval to use:
+
+        * ``'hpd'``: Calculate highest (average) posterior density (HPD)
+          interval, equivalent to iso-pdf interval (interval with same
+          probability density at each end, also known as waterline-interval).
+          This only works if the distribution is unimodal.
+        * ``'ll'``/``'ul'``: Lower/upper limit. One-sided limits for which
+          ``level`` fraction of the (equally weighted) samples lie above/below
+          the limit.
+        * ``'et'``: Equal-tailed interval with the same fraction of (equally
+          weighted) samples below and above the interval region.
+
+    u : array, optional
+        Random values to use for reproducible sample compression.
+        Default: ``np.random.rand(len(weights))`` if non-unit weights used.
```

Comment:
> Should the random value array u be part of this function? I feel like this is combining two things (sample compression and computation of the C.I.) into one function, where it would make more sense to handle them separately...?

The only reason for this to be here was to allow to get reproducible outputs, i.e. pass through the `u` argument from `compress_samples`. But I realize that `np.random.seed(0)` before the function call already does this so the `u` is not necessary; deleted.

Reply:
I agree that we don't have precise knowledge of the population. But I think we can make our results more robust to variability coming from the sample compression (now that we got rid of the `u`). If we move the line
```python 
weights = compress_weights(weights, ncompress=-1)
```
into our iteration loop, then each iteration will use a slightly different subsample, giving us a handle at the (sub-)sample variance (or maybe it is more accurate to call this the compression variance...?).